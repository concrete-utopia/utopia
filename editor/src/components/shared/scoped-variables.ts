import type { ElementPath } from '../../core/shared/project-file-types'
import { withUnderlyingTarget } from '../editor/store/editor-state'
import type { ProjectContentTreeRoot } from '../assets'
import type {
  ElementInstanceMetadataMap,
  ArbitraryJSBlock,
  TopLevelElement,
  UtopiaJSXComponent,
} from '../../core/shared/element-template'
import {
  isArbitraryJSBlock,
  isUtopiaJSXComponent,
  jsxAttributesFromMap,
  jsxElementWithoutUID,
  jsExpressionValue,
  jsxConditionalExpressionWithoutUID,
  emptyComments,
  jsExpressionOtherJavaScriptSimple,
  jsxFragmentWithoutUID,
  jsxTextBlock,
  isJSXElement,
  isJSXAttributesEntry,
  isJSXAttributeValue,
  simplifyAttributeIfPossible,
} from '../../core/shared/element-template'
import type { FileRootPath, VariableData } from '../canvas/ui-jsx-canvas'
import { type VariablesInScope } from '../canvas/ui-jsx-canvas'
import { getContainingComponent, toComponentId } from '../../core/shared/element-path'
import {
  type InsertableComponentGroup,
  type InsertableVariableType,
  insertableComponentGroupProjectComponent,
  insertableVariable,
} from './project-components'
import { emptyImports } from '../../core/workers/common/project-file-utils'
import { isImage } from '../../core/shared/utils'
import { type ComponentElementToInsert } from '../custom-code/code-file'
import { omitWithPredicate } from '../../core/shared/object-utils'
import { MetadataUtils } from '../../core/model/element-metadata-utils'
import { isLeft } from '../../core/shared/either'
import { findContainingComponent } from '../../core/model/element-template-utils'

export function getVariablesInScope(
  elementPath: ElementPath | null,
  projectContents: ProjectContentTreeRoot,
  variablesInScopeFromEditorState: VariablesInScope,
  jsxMetadata: ElementInstanceMetadataMap,
): AllVariablesInScope[] {
  return withUnderlyingTarget(elementPath, projectContents, [], (success) => {
    let varsInScope = []

    if (elementPath !== null) {
      const containingComponent = findContainingComponent(success.topLevelElements, elementPath)
      const componentScopedVariables = getVariablesFromComponent(
        containingComponent,
        elementPath,
        variablesInScopeFromEditorState,
      )
      varsInScope.push(componentScopedVariables)

      const componentPropsInScope = getComponentPropsInScope(
        containingComponent,
        elementPath,
        jsxMetadata,
      )
      varsInScope.push(componentPropsInScope)
    }

    /** for future reference - adding variables from top level **/
    // const topLevelVariables = getTopLevelVariables(success.topLevelElements, underlyingFilePath)
    // varsInScope.push(topLevelVariables)

    return varsInScope
  })
}

function getTopLevelVariables(topLevelElements: TopLevelElement[], underlyingFilePath: string) {
  const topLevelCode: ArbitraryJSBlock[] = topLevelElements.filter((el) =>
    isArbitraryJSBlock(el),
  ) as ArbitraryJSBlock[]
  const topLevelVariables = topLevelCode?.map((block) => block?.definedWithin ?? []).flat()

  return {
    filePath: underlyingFilePath,
    variables: topLevelVariables.map((variable) => ({
      name: variable,
      type: 'string' as const,
    })),
  }
}

function getVariablesFromComponent(
  jsxComponent: UtopiaJSXComponent | null,
  elementPath: ElementPath,
  variablesInScopeFromEditorState: VariablesInScope,
): AllVariablesInScope {
  const elementPathString = toComponentId(elementPath)
  const jsxComponentVariables = variablesInScopeFromEditorState[elementPathString] ?? {}
  return {
    filePath: jsxComponent?.name ?? 'Component',
    variables: generateVariableTypes(jsxComponentVariables),
  }
}

function getComponentPropsInScope(
  jsxComponent: UtopiaJSXComponent | null,
  elementPath: ElementPath,
  jsxMetadata: ElementInstanceMetadataMap,
): AllVariablesInScope {
  const jsxComponentPropNamesDeclared = jsxComponent?.propsUsed ?? []

  const jsxComponentPropsPassed = omitWithPredicate(
    getContainerPropsValue(elementPath, jsxMetadata),
    (key) => typeof key !== 'string' || !jsxComponentPropNamesDeclared.includes(key),
  )

  const name = jsxComponent?.name ?? 'Component'
  return {
    filePath: `${name}::props`,
    variables: generateVariableTypes(jsxComponentPropsPassed),
  }
}

function getContainerPropsValue(
  elementPath: ElementPath,
  jsxMetadata: ElementInstanceMetadataMap,
): VariableData {
  const containerMetadata = MetadataUtils.findElementByElementPath(
    jsxMetadata,
    getContainingComponent(elementPath),
  )
  const runtimeProps: VariableData = {}

  if (containerMetadata == null || isLeft(containerMetadata.element)) {
    return runtimeProps
  }

  const containerElement = containerMetadata.element.value

  if (isJSXElement(containerElement)) {
    containerElement.props.forEach((prop) => {
      if (isJSXAttributesEntry(prop)) {
        const value = simplifyAttributeIfPossible(prop.value)
        if (isJSXAttributeValue(value)) {
          runtimeProps[prop.key] = {
            spiedValue: value.value as unknown,
            insertionCeiling: elementPath,
          }
        }
      }
    })
  }

  return runtimeProps
}

function generateVariableTypes(variables: VariableData): Variable[] {
  return Object.entries(variables).flatMap(([name, valueMetadata]) => {
    const value = valueMetadata.spiedValue
    const insertionCeiling = valueMetadata.insertionCeiling
    const type = getTypeByValue(value)
    const variable: Variable = {
      name,
      value,
      type,
      depth: 0,
      insertionCeiling: insertionCeiling,
    }
    if (type === 'object' && value != null) {
      // iterate also first-level keys of object
      return [variable].concat(
        Object.entries(value).map(([key, innerValue]) => ({
          displayName: `${key}`,
          name: `${name}.${key}`,
          value: innerValue,
          type: getTypeByValue(innerValue),
          parent: variable,
          depth: 1,
          insertionCeiling: insertionCeiling,
        })),
      )
    } else {
      return variable
    }
  })
}

export function convertVariablesToElements(
  variableInScope: AllVariablesInScope[],
): InsertableComponentGroup[] {
  return variableInScope.map((variableGroup) => {
    return {
      source: insertableComponentGroupProjectComponent(variableGroup.filePath),
      insertableComponents: variableGroup.variables.map((variable) => {
        return insertableVariable(
          emptyImports(),
          () => getMatchingElementForVariable(variable),
          variable.displayName ?? variable.name,
          [],
          null,
          variable.type,
          variable.depth,
          variable.name,
          variable.insertionCeiling,
        )
      }),
    }
  })
}

function getMatchingElementForVariable(variable: Variable): ComponentElementToInsert {
  return getMatchingElementForVariableInner(variable).component
}

function getMatchingElementForVariableInner(variable: Variable): InsertableComponentAndJSX {
  const originalVariableName = variable.parent != null ? variable.parent.name : variable.name
  switch (variable.type) {
    case 'string':
      return simpleInsertableComponentAndJsx('span', variable.name, originalVariableName)
    case 'number':
      return simpleInsertableComponentAndJsx('span', variable.name, originalVariableName)
    case 'boolean':
      return insertableComponentAndJSX(
        jsxConditionalExpressionWithoutUID(
          jsExpressionOtherJavaScriptSimple(variable.name, [originalVariableName]),
          variable.name,
          jsExpressionValue(null, emptyComments),
          jsExpressionValue(null, emptyComments),
          emptyComments,
        ),
        `${variable.name} ? null : null`,
      )
    case 'object':
      return simpleInsertableComponentAndJsx(
        'span',
        `JSON.stringify(${variable.name})`,
        originalVariableName,
      )
    case 'array':
      return arrayInsertableComponentAndJsx(variable)
    case 'image':
      return insertableComponentAndJSX(
        jsxElementWithoutUID(
          'img',
          jsxAttributesFromMap({
            src: jsExpressionOtherJavaScriptSimple(variable.name, [originalVariableName]),
          }),
          [],
        ),
        `<img src={${variable.name}} style={{width: 100, height: 100, contain: 'layout'}}/>`,
      )
    default:
      return simpleInsertableComponentAndJsx(
        'span',
        `JSON.stringify(${variable.name})`,
        originalVariableName,
      )
  }
}

function getTypeByValue(value: unknown): InsertableVariableType {
  const type = typeof value
  if (type === 'object' && Array.isArray(value)) {
    return 'array'
  } else if (typeof value === 'string' && isImage(value)) {
    return 'image'
  }
  return type
}

function arrayInsertableComponentAndJsx(variable: Variable): InsertableComponentAndJSX {
  const arrayElementsType = getArrayType(variable)
  const innerElementString = getMatchingElementForVariableInner({
    name: 'item',
    type: arrayElementsType,
    depth: variable.depth + 1,
    insertionCeiling: variable.insertionCeiling,
  }).jsx
  const arrayElementString = `${variable.name}.map((item) => (${innerElementString}))`
  const arrayElement = jsxFragmentWithoutUID([jsxTextBlock(`{${arrayElementString}}`)], true)
  return insertableComponentAndJSX(arrayElement, arrayElementString)
}

function getArrayType(arrayVariable: Variable): InsertableVariableType {
  const arr = arrayVariable.value as unknown[]
  const types = new Set(arr.map((item) => getTypeByValue(item)))
  return types.size === 1 ? [...types][0] : 'object'
}

function insertableComponentAndJSX(
  component: ComponentElementToInsert,
  jsx: string,
): InsertableComponentAndJSX {
  return {
    component,
    jsx,
  }
}

function simpleInsertableComponentAndJsx(
  tag: string,
  expression: string,
  variableName: string = expression,
): InsertableComponentAndJSX {
  return insertableComponentAndJSX(
    jsxElementWithoutUID(tag, jsxAttributesFromMap({}), [
      jsExpressionOtherJavaScriptSimple(expression, [variableName]),
    ]),
    `<${tag}>{${expression}}</${tag}>`,
  )
}

interface InsertableComponentAndJSX {
  component: ComponentElementToInsert
  jsx: string
}

export type AllVariablesInScope = {
  filePath: string
  variables: Variable[]
}

interface Variable {
  name: string
  type: InsertableVariableType
  displayName?: string
  value?: unknown
  parent?: Variable
  depth: number
  insertionCeiling: FileRootPath | ElementPath
}
