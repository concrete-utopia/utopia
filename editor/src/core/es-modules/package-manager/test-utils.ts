import type { PackagerServerFileDescriptor } from '../../shared/npm-dependency-types'
import type { NodeModules } from '../../shared/project-file-types'
import { esCodeFile } from '../../shared/project-file-types'
import { objectMap } from '../../shared/object-utils'

export function createNodeModules(contents: {
  [filepath: string]: PackagerServerFileDescriptor
}): NodeModules {
  return objectMap(
    (content, key: string) => esCodeFile(content.content, 'NODE_MODULES', key),
    contents,
  )
}
