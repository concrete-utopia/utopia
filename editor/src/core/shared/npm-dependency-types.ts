export type RequireFn = (importOrigin: string, toImport: string) => any
export type TypeDefinitions = {
  [fileName: string]: string
} // the strings are the contents of .d.ts files

export type NpmDependency = {
  name: string
  version: string
}

export function npmDependency(name: string, version: string) {
  return {
    name: name,
    version: version,
  }
}

export interface PackagerServerFileDescriptor {
  content: string
}

export interface PackagerServerResponse {
  contents: {
    [filepath: string]: PackagerServerFileDescriptor
  }
}

export interface JsdelivrResponse {
  default: string
  files: Array<{
    name: string
    hash: string
    time: string
    size: number
  }>
}
