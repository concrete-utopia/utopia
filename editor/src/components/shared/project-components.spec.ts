import type {
  PackageStatusMap,
  PossiblyUnversionedNpmDependency,
} from '../../core/shared/npm-dependency-types'
import { resolvedNpmDependency } from '../../core/shared/npm-dependency-types'
import { DefaultThirdPartyControlDefinitions } from '../../core/third-party/third-party-controls'
import { simpleDefaultProjectPreParsed } from '../../sample-projects/sample-project-utils.test-utils'
import type { PropertyControlsInfo } from '../custom-code/code-file'
import {
  clearInsertableComponentGroupUniqueIDs,
  getComponentGroups,
  getDependencyStatus,
} from './project-components'

describe('getComponentGroups', () => {
  it('returns all the various default groups', () => {
    const packageStatus: PackageStatusMap = {
      antd: { status: 'loaded' },
    }
    const dependencies: Array<PossiblyUnversionedNpmDependency> = [
      resolvedNpmDependency('antd', '4.1.0'),
    ]

    const propertyControlsInfo: PropertyControlsInfo = {
      antd: DefaultThirdPartyControlDefinitions.antd,
    }
    const actualResult = getComponentGroups(
      'insert',
      packageStatus,
      propertyControlsInfo,
      simpleDefaultProjectPreParsed().projectContents,
      dependencies,
      '/src/app.js',
    ).map(clearInsertableComponentGroupUniqueIDs)

    expect(actualResult).toMatchSnapshot()
  })
  it('handles a sub-dependency correctly', () => {
    const packageStatus: PackageStatusMap = {
      '@heroicons/react': { status: 'loaded' },
    }
    const dependencies: Array<PossiblyUnversionedNpmDependency> = [
      resolvedNpmDependency('@heroicons/react', '1.0.5'),
    ]

    const propertyControlsInfo: PropertyControlsInfo = {
      '@heroicons/react/solid': {
        BeakerIcon: {
          properties: {},
          variants: [],
        },
      },
    }
    const actualResult = getComponentGroups(
      'insert',
      packageStatus,
      propertyControlsInfo,
      simpleDefaultProjectPreParsed().projectContents,
      dependencies,
      '/src/app.js',
    ).map(clearInsertableComponentGroupUniqueIDs)

    expect(actualResult).toMatchSnapshot()
  })
})

describe('getDependencyStatus', () => {
  it('shows the dependency as loaded when property controls exist for a sub-dependency', () => {
    const propertyControlsInfo: PropertyControlsInfo = {
      '@heroicons/react/solid': {
        BeakerIcon: {
          properties: {},
          variants: [],
        },
      },
    }
    const packageStatus: PackageStatusMap = {
      '@heroicons/react': {
        status: 'loaded',
      },
    }
    const result = getDependencyStatus(
      packageStatus,
      propertyControlsInfo,
      '@heroicons/react',
      'version-lookup',
    )
    expect(result).toEqual('loaded')
  })
})
