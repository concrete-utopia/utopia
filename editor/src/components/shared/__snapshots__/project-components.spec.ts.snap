// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getComponentGroups handles a sub-dependency correctly 1`] = `
Array [
  Object {
    "insertableComponents": Array [
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {},
        "insertionCeiling": null,
        "name": "App",
        "stylePropOptions": Array [
          "do-not-add",
        ],
      },
    ],
    "source": Object {
      "path": "/src/app.js",
      "type": "PROJECT_COMPONENT_GROUP",
    },
  },
  Object {
    "insertableComponents": Array [],
    "source": Object {
      "path": "/src/index.js",
      "type": "PROJECT_COMPONENT_GROUP",
    },
  },
  Object {
    "insertableComponents": Array [],
    "source": Object {
      "path": "/utopia/storyboard.js",
      "type": "PROJECT_COMPONENT_GROUP",
    },
  },
  Object {
    "insertableComponents": Array [
      Object {
        "defaultSize": Object {
          "height": 100,
          "width": 100,
        },
        "element": [Function],
        "importsToAdd": Object {
          "react": Object {
            "importedAs": "React",
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "div",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
    ],
    "source": Object {
      "type": "HTML_DIV",
    },
  },
  Object {
    "insertableComponents": Array [
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "react": Object {
            "importedAs": "React",
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "span",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "react": Object {
            "importedAs": "React",
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "h1",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "react": Object {
            "importedAs": "React",
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "h2",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "react": Object {
            "importedAs": "React",
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "p",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "react": Object {
            "importedAs": "React",
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "button",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "react": Object {
            "importedAs": "React",
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "input",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "react": Object {
            "importedAs": "React",
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "video",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "react": Object {
            "importedAs": "React",
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "img",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
    ],
    "source": Object {
      "type": "HTML_GROUP",
    },
  },
  Object {
    "insertableComponents": Array [
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {},
        "insertionCeiling": null,
        "name": "Conditional",
        "stylePropOptions": Array [
          "do-not-add",
        ],
      },
    ],
    "source": Object {
      "type": "CONDITIONALS_GROUP",
    },
  },
  Object {
    "insertableComponents": Array [
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "react": Object {
            "importedAs": "React",
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "Fragment",
        "stylePropOptions": Array [
          "do-not-add",
        ],
      },
    ],
    "source": Object {
      "type": "FRAGMENT_GROUP",
    },
  },
  Object {
    "insertableComponents": Array [
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {},
        "insertionCeiling": null,
        "name": "Sample text",
        "stylePropOptions": Array [
          "do-not-add",
        ],
      },
    ],
    "source": Object {
      "type": "SAMPLES_GROUP",
    },
  },
  Object {
    "insertableComponents": Array [],
    "source": Object {
      "dependencyName": "@heroicons/react",
      "dependencyStatus": "loaded",
      "type": "PROJECT_DEPENDENCY_GROUP",
    },
  },
]
`;

exports[`getComponentGroups returns all the various default groups 1`] = `
Array [
  Object {
    "insertableComponents": Array [
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {},
        "insertionCeiling": null,
        "name": "App",
        "stylePropOptions": Array [
          "do-not-add",
        ],
      },
    ],
    "source": Object {
      "path": "/src/app.js",
      "type": "PROJECT_COMPONENT_GROUP",
    },
  },
  Object {
    "insertableComponents": Array [],
    "source": Object {
      "path": "/src/index.js",
      "type": "PROJECT_COMPONENT_GROUP",
    },
  },
  Object {
    "insertableComponents": Array [],
    "source": Object {
      "path": "/utopia/storyboard.js",
      "type": "PROJECT_COMPONENT_GROUP",
    },
  },
  Object {
    "insertableComponents": Array [
      Object {
        "defaultSize": Object {
          "height": 100,
          "width": 100,
        },
        "element": [Function],
        "importsToAdd": Object {
          "react": Object {
            "importedAs": "React",
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "div",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
    ],
    "source": Object {
      "type": "HTML_DIV",
    },
  },
  Object {
    "insertableComponents": Array [
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "react": Object {
            "importedAs": "React",
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "span",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "react": Object {
            "importedAs": "React",
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "h1",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "react": Object {
            "importedAs": "React",
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "h2",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "react": Object {
            "importedAs": "React",
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "p",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "react": Object {
            "importedAs": "React",
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "button",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "react": Object {
            "importedAs": "React",
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "input",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "react": Object {
            "importedAs": "React",
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "video",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "react": Object {
            "importedAs": "React",
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "img",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
    ],
    "source": Object {
      "type": "HTML_GROUP",
    },
  },
  Object {
    "insertableComponents": Array [
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {},
        "insertionCeiling": null,
        "name": "Conditional",
        "stylePropOptions": Array [
          "do-not-add",
        ],
      },
    ],
    "source": Object {
      "type": "CONDITIONALS_GROUP",
    },
  },
  Object {
    "insertableComponents": Array [
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "react": Object {
            "importedAs": "React",
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "Fragment",
        "stylePropOptions": Array [
          "do-not-add",
        ],
      },
    ],
    "source": Object {
      "type": "FRAGMENT_GROUP",
    },
  },
  Object {
    "insertableComponents": Array [
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {},
        "insertionCeiling": null,
        "name": "Sample text",
        "stylePropOptions": Array [
          "do-not-add",
        ],
      },
    ],
    "source": Object {
      "type": "SAMPLES_GROUP",
    },
  },
  Object {
    "insertableComponents": Array [
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "antd": Object {
            "importedAs": null,
            "importedFromWithin": Array [
              Object {
                "alias": "DatePicker",
                "name": "DatePicker",
              },
            ],
            "importedWithName": null,
          },
          "antd/dist/antd.css": Object {
            "importedAs": null,
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "DatePicker",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "antd": Object {
            "importedAs": null,
            "importedFromWithin": Array [
              Object {
                "alias": "Button",
                "name": "Button",
              },
            ],
            "importedWithName": null,
          },
          "antd/dist/antd.css": Object {
            "importedAs": null,
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "Button",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "antd": Object {
            "importedAs": null,
            "importedFromWithin": Array [
              Object {
                "alias": "InputNumber",
                "name": "InputNumber",
              },
            ],
            "importedWithName": null,
          },
          "antd/dist/antd.css": Object {
            "importedAs": null,
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "InputNumber",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "antd": Object {
            "importedAs": null,
            "importedFromWithin": Array [
              Object {
                "alias": "Row",
                "name": "Row",
              },
            ],
            "importedWithName": null,
          },
          "antd/dist/antd.css": Object {
            "importedAs": null,
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "Row",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "antd": Object {
            "importedAs": null,
            "importedFromWithin": Array [
              Object {
                "alias": "Col",
                "name": "Col",
              },
            ],
            "importedWithName": null,
          },
          "antd/dist/antd.css": Object {
            "importedAs": null,
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "Col",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "antd": Object {
            "importedAs": null,
            "importedFromWithin": Array [
              Object {
                "alias": "Space",
                "name": "Space",
              },
            ],
            "importedWithName": null,
          },
          "antd/dist/antd.css": Object {
            "importedAs": null,
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "Space",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "antd": Object {
            "importedAs": null,
            "importedFromWithin": Array [
              Object {
                "alias": "Menu",
                "name": "Menu",
              },
            ],
            "importedWithName": null,
          },
          "antd/dist/antd.css": Object {
            "importedAs": null,
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "Menu",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "antd": Object {
            "importedAs": null,
            "importedFromWithin": Array [
              Object {
                "alias": "Menu",
                "name": "Menu",
              },
            ],
            "importedWithName": null,
          },
          "antd/dist/antd.css": Object {
            "importedAs": null,
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "Menu.Item",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "antd": Object {
            "importedAs": null,
            "importedFromWithin": Array [
              Object {
                "alias": "Menu",
                "name": "Menu",
              },
            ],
            "importedWithName": null,
          },
          "antd/dist/antd.css": Object {
            "importedAs": null,
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "Menu.SubMenu",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "antd": Object {
            "importedAs": null,
            "importedFromWithin": Array [
              Object {
                "alias": "Menu",
                "name": "Menu",
              },
            ],
            "importedWithName": null,
          },
          "antd/dist/antd.css": Object {
            "importedAs": null,
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "Menu.ItemGroup",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
      Object {
        "defaultSize": null,
        "element": [Function],
        "importsToAdd": Object {
          "antd": Object {
            "importedAs": null,
            "importedFromWithin": Array [
              Object {
                "alias": "Typography",
                "name": "Typography",
              },
            ],
            "importedWithName": null,
          },
          "antd/dist/antd.css": Object {
            "importedAs": null,
            "importedFromWithin": Array [],
            "importedWithName": null,
          },
        },
        "insertionCeiling": null,
        "name": "Typography.Text",
        "stylePropOptions": Array [
          "do-not-add",
          "add-size",
        ],
      },
    ],
    "source": Object {
      "dependencyName": "antd",
      "dependencyStatus": "loaded",
      "type": "PROJECT_DEPENDENCY_GROUP",
    },
  },
]
`;
